File: 728.自除数.py
#
# @lc app=leetcode.cn id=728 lang=python3
#
# [728] 自除数
#

# @lc code=start
class Solution:
    def selfDividingNumbers(self, left: int, right: int) -> List[int]:
        res = []
        l = list(range(left, right+1))
        flag = 0
        for num in l:
            s = str(num)
            for i in range(len(s)):
                if s[i] == '0' or num % int(s[i]) != 0:
                    flag = 1
                    break
                elif num % int(s[i]) == 0:
                    continue
            if flag == 0:
                res.append(num)
            flag = 0
        return res


        
# @lc code=end

File: 728.自除数.cpp
/*
 * @lc app=leetcode.cn id=728 lang=cpp
 *
 * [728] 自除数
 */

// @lc code=start
class Solution {
public:
    vector<int> selfDividingNumbers(int left, int right) {
        vector<int> res;
        while(left <= right){
            if(helper(left))
                res.push_back(left);
            left++;
        }
        return res;
    }
    bool helper(int val){
        int tmp = val;
        while(val != 0){
            if(val % 10 == 0)
                return false;
            if(tmp % (val % 10) == 0){
                val /= 10;
            }
            else 
                return false;
        }
        return true;
    }
};
// @lc code=end

